### ISSUE TITLE
Text search in pdf 

### ISSUE THREAD

**Issue by paveltech** [barteksc/AndroidPdfViewer#697]
Though i like this library very much. But i need a feature like search text in pdf and keyword will be highlighted and go to that page where keyword found.

**Comment by paveltech** [comment]
no still now i don't find this solution. If you found something let me know or if i found something i will comment.

**Comment by erfanfaramarzi** [comment]
Hi Masud
How are you?
Unfortunately I couldn't find any solution for searching through PDF
android app. I used a PDF viewer but it doesn't have search function.
I couldn't understand! Sorry! did you mean you could find a way to search
through PDF?

**Comment by Nourhanjx** [comment]
Yes!
What i meant to say that i’ll work to solve this issue and add the search feature into the application! 
But I guess you guys should approve first before i start the implementation of it :)?

**Comment by erfanfaramarzi** [comment]
I can't understand you.
Can you tell me how?

**Comment by erfanfaramarzi** [comment]
Would you mind please how you can search through a pdf in android???

>

**Comment by paveltech** [comment]
@Nourhanjx  how i can help you? and what do you mean by approve first? what need to approve.

**Comment by erfanfaramarzi** [comment]
I want to search through PDF file. Is it possible to tell me how

**Comment by hammadkhan** [comment]
> Yes!
> What i meant to say that i’ll work to solve this issue and add the search feature into the application!
> But I guess you guys should approve first before i start the implementation of it :)?

I don't think you need any approval to fork the code or implement your solution. You check check it out and make changes.

**Comment by kunang15** [comment]
how can we add a search in pdfview

**Comment by shivchand** [comment]
any one can help me to highlight text or paragraph while reading pdf ? Let me know how to code it

**Comment by Mukhtarali6121** [comment]
> Though i like this library very much. But i need a feature like search text in pdf and keyword will be highlighted and go to that page where keyword found.

Did u find any?

**Comment by lijusparkt** [comment]
Hlo,
Is it possible to Search a word and Indexing in this library...?

**Comment by sca0610** [comment]
This thread is still active. Hopefully someone can provide us solution for good. :+1:

**Comment by ncpy** [comment]
[look](https://stackoverflow.com/questions/66586673/how-to-implement-a-search-function-in-my-android-pdf-reader-app) here.

`implementation 'com.google.android.gms:play-services-vision:20.1.3' // search from pdf`

`ParcelFileDescriptor fileDescriptor2 = getContentResolver().openFileDescriptor(Uri.parse(getSource), "r");
 PdfRenderer pdfRenderer = new PdfRenderer(fileDescriptor2);

 for (int i=0; i<pdfRenderer.getPageCount(); i++) {
      PdfRenderer.Page page = pdfRenderer.openPage(i);
      Bitmap bmp = Bitmap.createBitmap(page.getWidth(), page.getHeight(), Bitmap.Config.ARGB_8888);
      page.render(bmp, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY);

      TextRecognizer textRecognizer = new TextRecognizer.Builder(getApplicationContext()).build();
      Frame imageFrame = new Frame.Builder().setBitmap(bmp).build();
      String imageText = "";
      SparseArray<TextBlock> textBlocks = textRecognizer.detect(imageFrame);
      
      for (int j = 0; j < textBlocks.size(); j++) {
            TextBlock textBlock = textBlocks.get(textBlocks.keyAt(j));
            imageText = textBlock.getValue();                   // return string
                        //Log.i("ders", imageText);
       }
       page.close();
 }
 pdfRenderer.close();`

**Comment by Sohumkalia3939** [comment]
help pls

**Comment by Attaroqqi** [comment]
@ncpy do you have source code to zoom the next textblocks when there are many detected bitmaps?

**Comment by vimalcvs** [comment]
This app is all ready done anyone can use reverse engineering.

https://play.google.com/store/apps/details?id=pdf.reader.pdfviewer.pdfeditor

**Comment by vimalcvs** [comment]
> Though i like this library very much. But i need a feature like search text in pdf and keyword will be highlighted and go to that page where keyword found.

This app is all ready done anyone can use reverse engineering.

https://play.google.com/store/apps/details?id=pdf.reader.pdfviewer.pdfeditor

**Comment by alizahid1996** [comment]
I have used [https://github.com/barteksc/AndroidPdfViewer.git](url) for Pdf Viewer and combined it with
**implementation 'com.tom-roush:pdfbox-android:2.0.27.0'** library to extract text from the pdf file

`
    private var matchingPages: List<Int> = emptyList()
    private var currentOccurrenceIndex = 0
    private var lastSearchedKeyword: String = ""
    private var filePath: String? = null


private fun searchPdf(etSearch: EditText) {
        val keyword = etSearch.text.toString()
       
        if (keyword != lastSearchedKeyword) {
            lastSearchedKeyword = keyword
            currentOccurrenceIndex = 0
            lifecycleScope.launch {
                filePath?.let {
                    matchingPages = findPagesForKeyword(it, keyword)
                    if (matchingPages.isEmpty()) {
                        showToast("Keyword not found")
                    } else {
                        binding.pdfView2.jumpTo(matchingPages[currentOccurrenceIndex], true)
                        currentOccurrenceIndex++
                    }
                  
                }
            }
        } else {
            if (matchingPages.isNotEmpty()) {
                if (currentOccurrenceIndex < matchingPages.size) {
                    binding.pdfView2.jumpTo(matchingPages[currentOccurrenceIndex], true)
                    currentOccurrenceIndex++
                } else {
                    showToast("No more occurrences")
                }
            } else {
                showToast("Keyword not found")
            }

         
      }
    }


//Function that finds Pages Index for the matched keyword and add the indexes in the list
private suspend fun findPagesForKeyword(pdfFilePath: String, keyword: String): List<Int> {
        return withContext(Dispatchers.IO) {
            val result = mutableListOf<Int>()
            val file = File(pdfFilePath)
            if (!file.exists()) return@withContext result

            PDDocument.load(file).use { document ->
                val stripper = PDFTextStripper()
                val totalPages = document.numberOfPages
                for (page in 1..totalPages) {
                    stripper.startPage = page
                    stripper.endPage = page
                    val pageText = stripper.getText(document)
                    if (pageText.contains(keyword, ignoreCase = true)) {
                        // Add the page index (zero-indexed)
                        result.add(page - 1)
                    }
                }
            }
            result
        }
    }

`

**Note: This code searches the text page by page and makes the pdf page jump to the page where the keyword is found**
